rustls/src/vecbuf.rs

```diff
@@ -200,8 +200,20 @@ impl ChunkVecBuffer {
             prefix = 0;
         }
         let len = cmp::min(bufs.len(), self.chunks.len());
-        let used = wr.write_vectored(&bufs[..len])?;
-        assert!(used <= self.len(), "illegal write_vectored return value");
+        let bufs = &bufs[..len];
+        let used = wr.write_vectored(bufs)?;
+        let available_bytes = bufs.iter().map(|ch| ch.len()).sum();
+
+        if used > available_bytes {
```

Comment:
maybe worth a link to https://github.com/rustls/rustls/issues/2316 here too? It feels like it's not immediately obvious why we check this, and the unit test that does have a ref is some distance away.